src/v3/src/util/theme.ts

```diff
@@ -10,116 +10,464 @@
+                borderColor: 'transparent',
+              },
+
+              "&:active": {
+                backgroundColor: mergedTokens.HueNeutral200,
+              },
+
+              "&:disabled": {
+                backgroundColor: "transparent",
+                color: mergedTokens.TypographyColorDisabled,
+              },
+            }),
+            ...(ownerState.size === "small" && {
+              paddingBlock: mergedTokens.Spacing2,
+              paddingInline: mergedTokens.Spacing3,
+              fontSize: mergedTokens.TypographySizeBody,
+            }),
+            ...(ownerState.size === "large" && {
+              paddingBlock: mergedTokens.Spacing4,
+              paddingInline: mergedTokens.Spacing4,
+            }),
+            ...(ownerState.fullWidth === true && {
+              display: "block",
+              width: "100%",
+              marginBlock: "0",
+              marginInline: "0",
+
+              "&:not(:last-child)": {
+                marginBlockEnd: mergedTokens.Spacing4,
+              },
+            }),
+            ...(ownerState.children === "" && {
+              minWidth: "auto",
+              padding: mergedTokens.Spacing3,
+
+              [`.${buttonClasses.endIcon}, .${buttonClasses.startIcon}`]: {
+                margin: "0",
+              },
+
+              ...(ownerState.size === "small" && {
+                padding: mergedTokens.Spacing2,
+              }),
+            }),
+          }),
+
+          endIcon: ({ ownerState }) => ({
+            display: "inline-flex",
+            margin: 0,
+            marginInlineStart: mergedTokens.Spacing2,
+
+            ...(ownerState.children === undefined && {
+              marginInlineStart: 0,
+            }),
+          }),
+
+          startIcon: ({ ownerState }) => ({
+            display: "inline-flex",
+            margin: 0,
+            marginInlineEnd: mergedTokens.Spacing2,
+
+            ...(ownerState.children === undefined && {
+              marginInlineEnd: 0,
+            }),
+          }),
+        },
+      },
+      MuiAlert: {
+        styleOverrides: {
+          root: {
+            gap: 0,
+          },
+          icon: ({ theme }) => ({
+            paddingInlineEnd: theme.spacing(4),
+            flexShrink: 0,
+          }),
+        },
+      },
+      MuiInputBase: {
+        styleOverrides: {
+          input: {
+            '::-ms-reveal': {
+              display: 'none',
+            },
+          },
+        },
+      },
+      // ruleset with :focus-visible pseudo-selector break entire ruleset in
+      // ie11 because its not supported. re-define the :hover rule separately
+      // again so the ruleset is applied in ie11
+      MuiIconButton: {
+        styleOverrides: {
+          root: () => ({
+            '&:hover': {
+              backgroundColor: 'rgba(29, 29, 33, 0.1)',
+              borderColor: 'transparent',
+            },
+          }),
+        },
+      },
+    }
+  }) as Theme;
```

Comment:
I wonder if this cast is still necessary. I have been hoping we can remove the casting in the code where possible because we lose all type safety

Reply:
Removed cast. Updated types